### ISSUE TITLE
Microsoft.Extensions.Logging v3.1.3 depends on DependencyInjection v3.1.0

### ISSUE THREAD

**Issue by SSBMTonberry** [dotnet/extensions#3213]
Hello!
In our project we recently got an error after upgrading Microsoft.Extensions.Logging and Microsoft.Extensions.DependencyInjection to the latest version (3.1.3). 

We got the following error:
```
ERROR : Error (0) Could not load file or assembly 'Microsoft.Extensions.DependencyInjection.Abstractions, Version=3.1.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
```
Which is strange, since we only refer to the v3.1.3 versions of all the Extension libraries, and have never really used the 3.1.0 version.

After some investigation, it seems like Microsoft.Extensions.Logging depends on an older version of 
Microsoft.Extensions.DependencyInjection:
![bilde](https://user-images.githubusercontent.com/26206421/80566194-b0d40f80-89f2-11ea-8867-1af1f5b2a9dc.png)

This is a part of a big solution where we use DependencyInjection for most underlying projects. 
The NuGet spec itself says it depends on v3.1.3, but the error and disassembly of the .dll suggests otherwise.

**Comment by anurse** [comment]
Are you running on .NET Framework? It's intentional that Logging references the `3.1.0` version of DI because the general policy is that it's safe to load a **later** version of an assembly than the one you referenced. This is true in .NET Core, where it will automatically accept (with no error) loading 3.1.3 when 3.1.0 was requested.

However, in .NET Framework, a [binding redirect](https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/file-schema/runtime/bindingredirect-element) is  needed to tell the framework that it's OK to load 3.1.3 even though 3.1.0 is requested. MSBuild should be automatically generating one for you though (see docs here: https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/how-to-enable-and-disable-automatic-binding-redirection).

Do you have a `TIFGeneratorService.exe.config` in the output of your `TIFGeneratorService` project containing `<bindingRedirect>` nodes? If not, something might be going wrong with the auto-generation.

**Comment by anurse** [comment]
Although, this looks like a duplicate of https://github.com/dotnet/extensions/issues/2931 .

**Comment by anurse** [comment]
Yep, closing this as a duplicate. Looks like the same problem.

**Comment by SSBMTonberry** [comment]
@anurse : Thanks for answering!
Yes, we are using .Net Framework 4.6.1, due to the application we are building on requires .Net Framework. For now I've chosen to simply downgrade all extension libs to 3.1.0, and that seems fine, but I am also generating a binding redirect in case I might need it for other .dlls.

At least I know the reason behind why it's done as it is, but I still believe the NuGet package for Extensions.Logging should state that it uses 3.1.0 or higher. It was actually pretty much hassle to figure out that this was the reason to my problems, and I wouldn't really expect a ".dll hell" since nothing really states that Extensions.Logging v3.1.3 depends on Extensions.DependencyInjection v3.1.0. And while this might not be an issue for .Net Core devs, it is an issue for people like me who don't have a choice in this case, but to use .Net Framework.

**Comment by anurse** [comment]
> At least I know the reason behind why it's done as it is, but I still believe the NuGet package for Extensions.Logging should state that it uses 3.1.0 or higher.

Yeah, it's not the clearest, but we're trying to strike a balance here. The way the package dependencies are expressed is what we'd prefer. You should, in general, try to keep your patch versions aligned (use the same `x` in `3.1.x` across the board). That's why the packages have the dependencies they have.

However, when you actually deploy to another machine, those packages might be part of .NET Core (ASP.NET Core, for example, embeds some of these packages so they won't end up in your publish output if you're building an ASP.NET Core app). In that situation, you might end up building against one patch version and then deploying to a **different** patch version. We want this to work, especially since it means you can roll out security fixes by just installing the patch on your server (without having to rebuild). That's why all the **assemblies** actually depend on the `3.1.0` version. We don't want a runtime binding failure just because you bumped up the version. Since .NET Core doesn't require binding redirects (it will load later versions of assemblies without complaint) that works well in that environment.

It does make .NET Framework usage a little more complicated. Unlike .NET Core, Framework requires exact version matches, or a binding redirect. That's an unfortunate trade-off but one we had to make in this situation.
